File: src/core/scheduler.py
import threading
import time
from datetime import datetime

import requests


class Scheduler:
    """HTTP request scheduler — runs requests on a repeating timer."""

    def __init__(self):
        self.is_running = False
        self._thread = None
        self._callbacks: dict = {}
        self.reset_stats()

    # ── Stats ──────────────────────────────────────────────────────────
    def reset_stats(self):
        self.request_count = 0
        self.success_count = 0
        self.total_ms = 0.0
        self.last_status = None

    @property
    def stats(self):
        total = self.request_count
        pct   = (self.success_count / total * 100) if total > 0 else 0.0
        avg   = (self.total_ms / total)             if total > 0 else 0.0
        return {
            "total":       total,
            "success":     self.success_count,
            "success_pct": pct,
            "avg_ms":      avg,
            "last_status": self.last_status,
        }

    # ── Event callbacks ─────────────────────────────────────────────────
    def on(self, event: str, callback):
        self._callbacks[event] = callback
        return self

    def _emit(self, event: str, *args, **kwargs):
        cb = self._callbacks.get(event)
        if cb:
            try:
                cb(*args, **kwargs)
            except Exception:
                pass

    # ── Control ──────────────────────────────────────────────────────────
    def start(self, url: str, method: str, headers: dict,
              body: str, interval: int, stop_time):
        if self.is_running:
            return
        self.reset_stats()
        self.is_running = True
        self._thread = threading.Thread(
            target=self._loop,
            args=(url, method, headers, body, interval, stop_time),
            daemon=True,
        )
        self._thread.start()

    def stop(self):
        self.is_running = False

    # ── Internal loop ────────────────────────────────────────────────────
    def _loop(self, url, method, headers, body, interval, stop_time):
        try:
            while self.is_running:
                if datetime.now().time() >= stop_time:
                    self._emit("completed", stop_time)
                    break

                self._make_request(url, method, headers, body)

                if not self.is_running:
                    break

                # tick-by-tick countdown
                for remaining in range(interval, 0, -1):
                    if not self.is_running:
                        break
                    self._emit("countdown", remaining, interval)
                    time.sleep(1)

        except Exception as e:
            self._emit("error_event", str(e))
        finally:
            self.is_running = False
            self._emit("finished")

    def _make_request(self, url, method, headers, body):
        t0 = time.time()
        ts = datetime.now().strftime("%H:%M:%S")
        try:
            resp = requests.request(
                method=method,
                url=url,
                headers=headers or {},
                data=body if body else None,
                timeout=10,
            )
            elapsed = (time.time() - t0) * 1000
            self.request_count += 1
            self.total_ms += elapsed
            code = resp.status_code
            if 200 <= code < 300:
                self.success_count += 1
            self.last_status = code
            self._emit("response", {
                "timestamp":  ts,
                "count":      self.request_count,
                "method":     method,
                "status":     code,
                "elapsed_ms": elapsed,
                "url":        url,
            })

        except requests.exceptions.Timeout:
            elapsed = (time.time() - t0) * 1000
            self.request_count += 1
            self.total_ms += elapsed
            self._emit("req_error", {"timestamp": ts, "count": self.request_count,
                                     "error": "Timed out (10 s)", "elapsed_ms": elapsed})

        except requests.exceptions.RequestException as e:
            elapsed = (time.time() - t0) * 1000
            self.request_count += 1
            self.total_ms += elapsed
            self._emit("req_error", {"timestamp": ts, "count": self.request_count,
                                     "error": str(e), "elapsed_ms": elapsed})

File: src-tauri/src/scheduler.rs
use std::sync::{Arc, atomic::{AtomicBool, Ordering}};
use tokio::time::{sleep, Duration};
use chrono::{Local, NaiveTime};
use tauri::{AppHandle, Emitter};
use crate::http_client::{make_request, RequestResult};
use crate::types::{ScheduleConfig, CountdownPayload, SentPayload, StatsPayload};

pub struct SchedulerHandle {
    stop_flag: Option<Arc<AtomicBool>>,
}

impl Default for SchedulerHandle {
    fn default() -> Self {
        Self { stop_flag: None }
    }
}

impl SchedulerHandle {
    pub fn is_running(&self) -> bool {
        self.stop_flag.as_ref().map_or(false, |f| f.load(Ordering::SeqCst))
    }

    pub fn start(&mut self, config: ScheduleConfig, client: reqwest::Client, app: AppHandle) -> Result<(), String> {
        if self.is_running() {
            return Err("Scheduler already running".into());
        }
        let flag = Arc::new(AtomicBool::new(true));
        self.stop_flag = Some(flag.clone());
        tokio::spawn(async move { run_scheduler(config, client, app, flag).await; });
        Ok(())
    }

    pub fn stop(&mut self) {
        if let Some(f) = &self.stop_flag { f.store(false, Ordering::SeqCst); }
    }
}

async fn run_scheduler(cfg: ScheduleConfig, client: reqwest::Client, app: AppHandle, running: Arc<AtomicBool>) {
    let mut total = 0u32;
    let mut success = 0u32;
    let mut total_ms = 0.0f64;
    let mut last_status: Option<u16> = None;

    let emit_stats = |t: u32, s: u32, ms: f64, ls: Option<u16>| {
        let _ = app.emit("scheduler:stats", StatsPayload {
            total: t, success: s,
            success_pct: if t > 0 { s as f64 / t as f64 * 100.0 } else { 0.0 },
            avg_ms: if t > 0 { ms / t as f64 } else { 0.0 },
            last_status: ls,
        });
    };

    if cfg.mode == "single" {
        total += 1;
        let res = make_request(&client, &cfg.url, &cfg.method, &cfg.headers, &cfg.body, total).await;
        handle_result(res, &mut success, &mut total_ms, &mut last_status, &app);
        emit_stats(total, success, total_ms, last_status);
        let _ = app.emit("scheduler:completed_single", ());
        running.store(false, Ordering::SeqCst);
        let _ = app.emit("scheduler:finished", ());
        return;
    }

    // Wait until start_time if set (continuous mode only)
    if cfg.mode == "continuous" && !cfg.start_time.is_empty() {
        if let Ok(st) = NaiveTime::parse_from_str(&cfg.start_time, "%H:%M") {
            while running.load(Ordering::SeqCst) {
                if Local::now().time() >= st { break; }
                let _ = app.emit("scheduler:waiting", cfg.start_time.clone());
                sleep(Duration::from_secs(1)).await;
            }
        }
    }

    let stop_time = if cfg.mode == "continuous" && !cfg.stop_time.is_empty() {
        NaiveTime::parse_from_str(&cfg.stop_time, "%H:%M").ok()
    } else { None };
    let mut sent_count = 0u32;

    while running.load(Ordering::SeqCst) {
        if cfg.mode == "continuous" {
            if let Some(st) = stop_time {
                if Local::now().time() >= st {
                    let _ = app.emit("scheduler:completed", cfg.stop_time.clone());
                    break;
                }
            }
        }
        if cfg.mode == "count" && sent_count >= cfg.count {
            let _ = app.emit("scheduler:completed_count", sent_count);
            break;
        }

        total += 1;
        let res = make_request(&client, &cfg.url, &cfg.method, &cfg.headers, &cfg.body, total).await;
        handle_result(res, &mut success, &mut total_ms, &mut last_status, &app);
        emit_stats(total, success, total_ms, last_status);
        sent_count += 1;
        let target = if cfg.mode == "count" { Some(cfg.count) } else { None };
        let _ = app.emit("scheduler:sent", SentPayload { count: sent_count, target });

        if cfg.mode == "count" && sent_count >= cfg.count {
            let _ = app.emit("scheduler:completed_count", sent_count);
            break;
        }
        if !running.load(Ordering::SeqCst) { break; }

        for rem in (1..=cfg.interval).rev() {
            if !running.load(Ordering::SeqCst) { break; }
            let _ = app.emit("scheduler:countdown", CountdownPayload { remaining: rem, total: cfg.interval });
            sleep(Duration::from_secs(1)).await;
        }
    }

    running.store(false, Ordering::SeqCst);
    let _ = app.emit("scheduler:finished", ());
}

fn handle_result(res: RequestResult, success: &mut u32, total_ms: &mut f64, last_status: &mut Option<u16>, app: &AppHandle) {
    match res {
        RequestResult::Success(d) => {
            if d.status >= 200 && d.status < 300 { *success += 1; }
            *total_ms += d.elapsed_ms;
            *last_status = Some(d.status);
            let _ = app.emit("scheduler:response", &d);
        }
        RequestResult::Failure(d) => {
            *total_ms += d.elapsed_ms;
            let _ = app.emit("scheduler:req_error", &d);
        }
    }
}
